### PR TITLE
Support handshake alerts

### PR DESCRIPTION
**Pull request by mickel8**
related with membraneframework/ex_dtls#8

### CODE REVIEW

File: lib/membrane_ice_plugin/handshake/handshake.ex [hunk 1]
```diff
@@ -26,51 +26,38 @@ defmodule Membrane.ICE.Handshake do
...
-  @callback recv_from_peer(state :: state(), data :: binary()) ::
+  @callback process(data :: binary(), state :: state()) ::
               :ok
               | {:ok, packets :: binary()}
-              | {:finished, handshake_data :: term(), packets :: binary()}
-              | {:finished, handshake_data :: term()}
+              | {:hsk_packets, packets :: binary()}
+              | {:hsk_finished, hsk_data :: any()}
+              | {:hsk_finished, hsk_data :: any(), packets :: binary()}
+              | {:error, value :: integer()}
```
Comment (mat-hek) [inline lib/membrane_ice_plugin/handshake/handshake.ex:57]: We're ignoring these errors for now, but is there any way we can handle them in the future?  Why is the value integer?

File: lib/membrane_ice_plugin/handshake/handshake.ex [hunk 2]
```diff
@@ -26,51 +26,38 @@ defmodule Membrane.ICE.Handshake do
...
+  If there is a need to retransmit some data send message `{:retransmit, id, data}`.
+  `id` is id assigned by ICE in `init/3` function.
   """
-  @callback recv_from_peer(state :: state(), data :: binary()) ::
+  @callback process(data :: binary(), state :: state()) ::
               :ok
               | {:ok, packets :: binary()}
-              | {:finished, handshake_data :: term(), packets :: binary()}
-              | {:finished, handshake_data :: term()}
+              | {:hsk_packets, packets :: binary()}
```
Comment (mat-hek) [inline lib/membrane_ice_plugin/handshake/handshake.ex:54]: I can understand abbreviating handshake to hsk in private variables, though I'd definitely avoid that in public API. Though that may be confusing to have handshake in some places and hsk in others.
P.s. please do such refactor in separate commits if possible ;)

File: lib/membrane_ice_plugin/sink.ex [hunk 3]
```diff
@@ -77,20 +72,28 @@ defmodule Membrane.ICE.Sink do
...
-    if Map.has_key?(ctx.pads, pad) do
-      {{:ok, get_initial_actions(pad, handshake_data)}, state}
+  defp maybe_send_demands(component_id, ctx, state) do
+    pad = Pad.ref(:input, component_id)
+    # if something is linked, component is ready and handshake is done then send demands
+    if Map.has_key?(ctx.pads, pad) and MapSet.member?(state.ready_components, component_id) and
+         Map.has_key?(state.finished_hsk, component_id) do
+      hsk_data = Map.get(state.finished_hsk, component_id)
+      actions = [demand: pad, event: {pad, %Handshake.Event{hsk_data: hsk_data}}]
+      {{:ok, actions}, state}
```
Comment (mat-hek) [inline lib/membrane_ice_plugin/sink.ex:94]: that's great, looks much better now ;)

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: lib/membrane_ice_plugin/handshake/handshake.ex (hunk 1 above)
Comment (mickel8) [reply lib/membrane_ice_plugin/handshake/handshake.ex:57]: I have changed it in ex_dtls so it returns `{:error, :peer_closed_for_writing}` at this moment. It's a good question how to handle such error. Should we stop ICE Bin or send this message to pipeline?

Comment (mat-hek) [reply lib/membrane_ice_plugin/handshake/handshake.ex:57]: I think in such cases we should raise ASAP. If one day we know how to handle that, we'll update the API

Comment (mickel8) [reply lib/membrane_ice_plugin/handshake/handshake.ex]: I am thinking about videroom. If we raise here our videoroom will fail so at the end we will still raise when DTLS alert occurs but with different message. Maybe let's convey this error up to the parent as `{:connection_closed, cause}`. In such scenario WebRTC plugin or videoroom can stop themselves.

Comment (mat-hek) [reply lib/membrane_ice_plugin/handshake/handshake.ex]: connection_closed shouldn't be returned as an error, it's rather a normal situation

Comment (mickel8) [reply lib/membrane_ice_plugin/handshake/handshake.ex]: True that, now we distinguish between error and closing connection. When error occurs we stop gen server.

Comment (mat-hek) [reply lib/membrane_ice_plugin/handshake/handshake.ex]: What about just raising in the handshake module, instead of returning error?

**Comment by mickel8** [comment]
I have also updated ex_libnice to 0.3.0

**Review by mat-hek — APPROVED** [review]

**Pull request merged**